text/0000-pack-phase-debug-script.md

```diff
@@ -0,0 +1,114 @@
+# Meta
+[meta]: #meta
+- Name: Pack phase debug script
+- Start Date: 2025-04-04
+- Author(s): BarDweller
+- Status: Draft 
+- RFC Pull Request:
+- CNB Pull Request: 
+- CNB Issue:
+- Supersedes: N/A
+
+# Summary
+[summary]: #summary
+
+An ability to dynamically add and invoke a debug script instead of the lifecycle binary per phase during `pack build`.
+
+# Motivation
+[motivation]: #motivation
+
+- Why should we do this?
+    - Debugging state within a builder can be tricky, it can be beneficial during buildpack and platform development to understand the content of buildpack toml files, and filesystem permissions, and even know sha hashes for builder packaged content. While it is possible to test some things via a custom debug buildpack, buildpacks are not involved in every phase of execution, and some changes occur outside of their involvement. Being able to trace those state changes and verify assumptions is beneficial to buildpack and platform developers alike. 
+- What use cases does it support?
+    - Example scenarios: 
+      - Trying to validate filesystem ownership/permissions/layout within a build container
+      - Validating cnb toml content pre-post phase execution to ensure changes are as expected
+      - Dumping checksum hashes of builder content to validate binaries being tested are the right ones
+
+# How it Works
+[how-it-works]: #how-it-works
+
+The pack cli would allow specification of a debug script, that should be driven instead of the appropriate lifecycle binary for each phase. 
+
+The script will be passed the name/path and arguments of the lifecycle binary that would have been driven. 
+
+The script is able to run abitrary commands and dump information to stdout/stderr, these commands can include invoking the original lifecycle binary with its original arguments. 
+
+A simple debug script could look like.. 
+
+```bash
+      #!/bin/bash
+      echo "Layer dir Content Before"
+      ls -lar /layers      
+      LIFECYCLE=$1
+      shift
+      $LIFECYCLE "$@"
+      echo "Layer dir Content After"
+      ls -lar /layers  
+```
+
+This dumps the content of the /layers dir, invokes the orignal command and then dumps it again.. 
+
+From a technical perspective, the debug script will be added to the ephemeral builder image at /cnb/lifecycle/debug, and used as the command when invoking the phase container. 
+
+# Migration
+[migration]: #migration
+
+This is a new feature, no migration issues are anticipated.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Adds clutter to an already complex set of commandline options for pack
```

Comment:
Would an env variable be less clutter? You could set something like `CNB_DEBUG_SCRIPT=...` and if set, it would invoke this behavior. You still have to document it, but then `pack build` doesn't need another arg.

Reply:
Env var works too.. although I'm not sure how many env vars are used to control pack behavior. Otomh, maybe DOCKER_HOST ?

Reply:
No strong feelings about one over the other. I agree that arguments would be more consistent with everything else in `pack build`.